pinot-core/src/main/java/org/apache/pinot/core/segment/memory/PinotNativeUnsafeByteBuffer.java

```diff
@@ -0,0 +1,355 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.segment.memory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.lang.reflect.Field;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.pinot.core.util.CleanerUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import sun.misc.Unsafe;
+import sun.nio.ch.DirectBuffer;
+
+
+@ThreadSafe
+public class PinotNativeUnsafeByteBuffer extends PinotDataBuffer {
+  private final DirectBuffer _buffer;
+  private final long _address;
+  private final boolean _flushable;
+  private static final Unsafe UNSAFE = getUnsafe();
+
+  private static Unsafe getUnsafe() {
+    try {
+      Field var0 = Unsafe.class.getDeclaredField("theUnsafe");
+      var0.setAccessible(true);
+      return (Unsafe)(var0.get((Object)null));
+    } catch (NoSuchFieldException e) {
+      throw new IllegalStateException("sun.misc.Unsafe is not available in this JVM", e);
+    } catch (IllegalAccessException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  public static PinotNativeUnsafeByteBuffer allocateDirect(int size) {
+    return new PinotNativeUnsafeByteBuffer(ByteBuffer.allocateDirect(size), true, false);
+  }
+
+  static PinotNativeUnsafeByteBuffer loadFile(File file, long offset, int size)
+      throws IOException {
+    PinotNativeUnsafeByteBuffer buffer = allocateDirect(size);
+    buffer.readFrom(0, file, offset, size);
+    return buffer;
+  }
+
+  static PinotNativeUnsafeByteBuffer mapFile(File file, boolean readOnly, long offset, int size)
+      throws IOException {
+    if (readOnly) {
+      try (FileChannel fileChannel = new RandomAccessFile(file, "r").getChannel()) {
```

Comment:
Nit: you can just have the read/write mode assigned based on readOnly flag, and still have the `fileChannel.map()` and `new PinotNativeUnsafeByteBuffer()` outside the if check.